File: Data Structures & Algorithms/majority-element/submission-0.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        cand = nums[0]
        cnt = 1

        for n in nums[1:]:
            if cnt == -1:
                cnt = 0
                cand = n
            if n == cand:
                cnt += 1
            else:
                cnt -= 1
        
        return cand

File: Data Structures & Algorithms/minimum-number-of-operations-to-make-array-empty/submission-0.cpp
/*
x%3: 1
x//3 - 1 + 2
1: -1
2: 1
3: 1
4: 2
5: 2
6: 2
7: 3
*/
class Solution {
public:
    int minOperations(vector<int>& nums) {
        unordered_map<int, int> numCnt;
        for (int n: nums) {
            numCnt[n]++;
        }

        int minOps = 0;
        for (auto [key, value]: numCnt) {
            if (numCnt[key] == 1) {
                return -1;
            }
            if (numCnt[key] % 3 == 0) {
                minOps += numCnt[key]/3;
            } else {
                minOps += (numCnt[key]/3 + 1);
            }
        }
        return minOps; 
    }
};